File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.py
def timeToMin(time):
    hh,mm = map(int,time.split(":"))
    return hh*60+mm

## 현재 시간과 다음 과제 꼭 시작해야 하는 시간(=gap) 구함
    ## gap 내에 현재 과제 다 할 수 있는 경우 => 정답에 추가
    ## gap 내 현재 과제 다 못하는 경우 => 남은 시간 갱신 해 

def solution(plans):
    answer = []
    stack = []
    plans.sort(key=lambda x:x[1]) ## 과제 꼭 시작해야 하는 순서대로 나열
    
    for idx in range(0, len(plans)-1):
        curr_, next_ = plans[idx], plans[idx+1]
        gap = timeToMin(next_[1])- timeToMin(curr_[1])
        stack.append([curr_[0], int(curr_[2])])
        while stack and gap:
            if stack[-1][-1]<=gap: #
                name, time = stack.pop()
                answer.append(name)
                gap-=time
            else:
                stack[-1][-1]-=gap
                gap=0
    answer.append(plans[idx+1][0])
    while stack:
        name, time = stack.pop()
        answer.append(name)


        
    return answer

File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.js
function formatMM(time){
    const [hh,mm]=time.split(":").map(Number)
    return hh*60+mm
}


function parseArr(plans){
    const arr = []
    for (let [a,b,c] of plans){
        arr.push([a,formatMM(b),Number(c)])
    }
    arr.sort((a,b)=>a[1]-b[1])
    return arr
}

function solution(plans) {
    plans = parseArr(plans)
    const answer = []
    const waits=[]
    let gapTime
    for(let i=0;i<plans.length-1;i++){
        const curr_ = plans[i]
        const next_ = plans[i+1]
        gapTime=next_[1]-curr_[1]
        waits.push([curr_[0], curr_[2]])
        while(waits.length && gapTime){
            
            if(waits[waits.length-1][1]-gapTime>0){
                waits[waits.length-1][1]-=gapTime
                gapTime=0
            }else{
                const item = waits.pop()
                answer.push(item[0])
                gapTime-=item[1]
            }
            
        }


    }
    answer.push(plans[plans.length-1][0])
    while(waits.length){
        answer.push(waits.pop()[0])
    }

    return answer;
}